packages/belos/tpetra/src/BelosTpetraOperator.hpp

```diff
@@ -0,0 +1,205 @@
+    /// The solver is initialized with a LinearProblem (\c lp) and ParameterList (\c plist).
+    ///
+    /// One should create a new linear problem to pass to the inner solver which is separate
+    /// from the linear problem passed to any outer solver. (Both problems may have the same 
+    /// operator, but solution and right-hand side vectors need to be maintained separately.)  
+    ///
+    /// When true, the bool \c initSolnVec causes the solution vector \c Y to be itialized to zero
+    /// when apply(X,Y) is called.
+
+  TpetraOperator( const Teuchos::RCP<LinearProblem<Scalar,Tpetra_MultiVector,Tpetra_Operator> >& lp, 
+		  const Teuchos::RCP<Teuchos::ParameterList>& plist,
+      const std::string solver_name,
+      bool initSolnVec = false )  
+  : lp_(lp), 
+    plist_(plist),
+    solver_name_(solver_name),
+    initSolnVec_(initSolnVec)
+  {
+
+    //  
+    // Create instance of solver. 
+    //  
+    Belos::SolverFactory<Scalar,Tpetra_MultiVector,Tpetra_Operator> factory;
+    solver_ = factory.create( solver_name_, plist );
+    solver_->setProblem( lp_ );
+
+    // Create a label for this Tpetra_Operator.
+    solver_name_ = "Belos " + solver_name_ + " Solver";
+
+    // Copy std::string to character array.  
+    // Not using conversion routine copy() because it's not supported by RW on Janus. (HKT 11/13/2003) 
+    Solver.resize(solver_name_.length()+1);
+    for (int i=0; i<(int)solver_name_.length(); i++) {
+      Solver[i] = solver_name_[i];
+    }   
+    Solver[solver_name_.length()] = 0;
+
+  }
+
+  
+  //! Destructor
+  virtual ~TpetraOperator() = default;
+  //@}
+  
+  //! @name Operator application methods
+  //@{ 
+  
+  /// \brief Apply the operator.  Performs a linear solve with right-hand side \c X.
+  /// Solution is stored in \c Y.  If \c initSolnVec was set to true, the vector \c Y
+  /// will be initialized with zeros before calling solve.  If \c initSolnVec is false, 
+  /// the vector \c Y will be used as an initial guess for the linear solve.  
+  ///
+  /// When using the GmresPoly solver, the polynomial will be applied to \c X at each call
+  /// instead of performing a linear solve.  
+
+  void apply( const Tpetra_MultiVector &X, Tpetra_MultiVector &Y, 
+           Teuchos::ETransp mode = Teuchos::NO_TRANS,
+           Scalar alpha = Teuchos::ScalarTraits<Scalar>::one(),
+           Scalar beta = Teuchos::ScalarTraits<Scalar>::zero()) const override
+  {
+    Teuchos::RCP<const Tpetra_MultiVector> vec_X = Teuchos::rcp( &X, false );
+    Teuchos::RCP<Tpetra_MultiVector> vec_Y = Teuchos::rcp( &Y, false );
+    if (initSolnVec_)
+    {
+      vec_Y->putScalar( 0.0 );
+      lp_->setInitResVec( vec_X );
+    }
+    lp_->setProblem( vec_Y, vec_X );
+    Belos::ReturnType ret = solver_->solve();
+  };
+  //@}
+  
+  //! @name Attribute access functions
+  //@{ 
+
+  //! Return the name of the requested solver.
+  const char* getSolverName() const { return Solver.data(); };
```

Comment:
1. Users would expect to be able to use this name in `Belos::SolverFactory`, to get an instance of the same solver used here.  However, users would be wrong -- you changed the original name they provided by adding some label stuff to it.  It would be better to give back either the original name, or the canonical name from `Belos::SolverFactory`.
2. You don't need the `std::vector<char>`.  Just return `solver_name_.c_str()`.
3. Why not return a `std::string`?

Reply:
@mhoemmen @jennloe The joys of inheritance!  Please eliminate the usage of char* in the Tpetra implementation of the operator.  The Label() function in Epetra returns const char*, which is the reason for character string manipulation including the addition of a null-termination character that was for Janus (a computer from 2003 with some bad juju compilers).